Source/ASDisplayNode.mm

```diff
@@ -2132,21 +2169,24 @@ - (void)recursivelyEnsureDisplaySynchronously:(BOOL)synchronously
 
 - (void)setShouldBypassEnsureDisplay:(BOOL)shouldBypassEnsureDisplay
 {
+  ASDN::MutexLocker l(__instanceLock__);
   _flags.shouldBypassEnsureDisplay = shouldBypassEnsureDisplay;
 }
 
 - (BOOL)shouldBypassEnsureDisplay
 {
+  ASDN::MutexLocker l(__instanceLock__);
   return _flags.shouldBypassEnsureDisplay;
 }
 
 - (void)setNeedsDisplayAtScale:(CGFloat)contentsScale
 {
-  ASDN::MutexLocker l(__instanceLock__);
-  if (contentsScale != self.contentsScaleForDisplay) {
-    self.contentsScaleForDisplay = contentsScale;
-    [self setNeedsDisplay];
+  if (contentsScale == self.contentsScaleForDisplay) {
+    return;
   }
+  
+  self.contentsScaleForDisplay = contentsScale;
```

Comment:
It's tempting to grab the lock here and drop down to `_contentsScaleForDisplay`.